crates/services/importer/src/importer.rs

```diff
@@ -112,17 +110,21 @@ impl PartialEq for Error {
     }
 }
 
+type ResultReceiver = Arc<Mutex<mpsc::Receiver<()>>>;
+
 pub struct Importer<D, E, V> {
     database: Mutex<D>,
     executor: Arc<E>,
     verifier: Arc<V>,
     chain_id: ChainId,
     broadcast: broadcast::Sender<ImporterResult>,
-    /// The channel to notify about the end of the processing of the previous block by all listeners.
-    /// It is used to await until all receivers of the notification process the `SharedImportResult`
-    /// before starting committing a new block.
-    prev_block_process_result: Mutex<Option<oneshot::Receiver<()>>>,
     guard: tokio::sync::Semaphore,
+    /// The `mpsc` channel is used to notify about the end of the processing of
+    /// the previous `SharedImportResult`. The importer waits until all receivers
+    /// of the notification process  the `SharedImportResult` before starting
+    /// committing a new block.
+    prev_block_process_result_sender: mpsc::Sender<()>,
+    prev_block_process_result_receiver: ResultReceiver,
```

Comment:
A better data type for limiting concurrency like this would be [Semaphore](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html), which is specifically the type to use to limit concurrency. Instead of holding `mpsc::Sender`, the processing party (i.e. `Awaiter`) could hold [SemaphorePermit](https://docs.rs/tokio/latest/tokio/sync/struct.SemaphorePermit.html), which automatically does the correct thing on deallocation.